feat: add C ndarray interface and refactor implementation for `stats/base/dvarmpn`

DhruvArvindSingh opened the pull request:
Resolves None.

## Description

> What is the purpose of this pull request?

This pull request:

-   adds C ndarray interface for `stats/base/dvarmpn`
-   refactors JavaScript implementation
-   updates tests, docs, benchmarks and examples

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves None 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

DhruvArvindSingh committed:
feat: added ndarray implementation for @stats/base/dvarmpn

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

DhruvArvindSingh committed:
refactor: refactored benchmarks in @stdlib/stats/base/dvarmpn

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: failed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: failed
---

DhruvArvindSingh committed:
fix: benchmarks

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: passed
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: failed
---

DhruvArvindSingh committed:
fix: benchmark.c

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

DhruvArvindSingh committed:
refactor: test and repl

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: passed
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: passed
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/examples/c/example.c:
```diff
@@ -17,21 +17,20 @@
 
-	// Specify the stride length:
-	int64_t stride = 2;
+	// Specify the strideX length:
```
```suggestion
	// Specify the stride length:
```

DhruvArvindSingh replied:
Resolved

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/benchmark/benchmark.js:
we use `@stdlib/array/linspace` only in packages which require a sorted array such as `stats/base/dmaxsorted` etc. can you please update the JavaScript benchmarks to use `@stdlib/random/array/uniform`

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/benchmark/benchmark.js:
```diff
@@ -39,13 +45,7 @@ var dvarmpn = require( './../lib/dvarmpn.js' );
-	for ( i = 0; i < x.length; i++ ) {
-		x[ i ] = ( randu()*20.0 ) - 10.0;
-	}
+	var x = linspace( -len/2, len/2, len, options );
```
```suggestion
	var x = uniform( len, -10.0, 10.0, options );
```
basically this.

you need to import `@stdlib/random/array/uniform`.

DhruvArvindSingh replied:
Resolved

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/benchmark/c/benchmark.length.c:
```diff
@@ -120,6 +121,40 @@ static double benchmark( int iterations, int len ) {
+	double t;
+	int i;
+
+	for ( i = 0; i < len; i++ ) {
+		x[ i ] = i;
+	}
```
```suggestion
	for ( i = 0; i < len; i++ ) {
		x[ i ] = ( rand_double() * 20000.0 ) - 10000.0;
	}
```
we don't need a sorted array here, so better to follow the same convention as the non-offset interface

DhruvArvindSingh replied:
Resolved

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/docs/repl.txt:
```diff
@@ -34,8 +34,8 @@
-    stride: integer
-        Index increment.
+    strideX: integer
+        Stride Length.
```
```suggestion
        Stride length.
```

DhruvArvindSingh replied:
Resolved

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/docs/repl.txt:
```diff
@@ -94,10 +93,10 @@
-    stride: integer
-        Index increment.
+    strideX: integer
+        Stride Length.
```
```suggestion
        Stride length.
```

DhruvArvindSingh replied:
Resolved

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/docs/types/index.d.ts:
```diff
@@ -29,7 +29,7 @@ interface Routine {
 	* @param correction - degrees of freedom adjustment
 	* @param x - input array
-	* @param stride - stride length
+	* @param strideX - strideX length
```
```suggestion
	* @param strideX - stride length
```

aayush0325 replied:
worth looking across the entire package for such search and replace errors

DhruvArvindSingh replied:
Resolved

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/docs/types/index.d.ts:
```diff
@@ -49,8 +49,8 @@ interface Routine {
 	* @param x - input array
-	* @param stride - stride length
-	* @param offset - starting index
+	* @param strideX - strideX length
```
```suggestion
	* @param strideX - stride length
```

DhruvArvindSingh replied:
Resolved

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/docs/types/index.d.ts:
```diff
@@ -71,7 +71,7 @@ interface Routine {
 * @param correction - degrees of freedom adjustment
 * @param x - input array
-* @param stride - stride length
+* @param strideX - strideX length
```
```suggestion
* @param strideX - stride length
```

DhruvArvindSingh replied:
Resolved

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/examples/index.js:
```diff
@@ -18,18 +18,13 @@
+var options = {
+	'dtype': 'float64'
+};
+var x = linspace( -5.0, 5.0, 10, options );
```
kindly use `@stdlib/random/array/discrete-uniform` here.

DhruvArvindSingh replied:
Resolved

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/include/stdlib/stats/base/dvarmpn.h:
```diff
@@ -31,7 +31,11 @@ extern "C" {
-double stdlib_strided_dvarmpn( const int64_t N, const double mean, const double correction, const double *X, const int64_t stride );
+double API_SUFFIX( stdlib_strided_dvarmpn )( const CBLAS_INT N, const double mean, const double correction, const double *X, const CBLAS_INT strideX );
+/**
+* Computes the variance of a double-precision floating-point strided array provided a known mean and using Neely's correction algorithm using alternative indexing semantics.
```
```suggestion
* Computes the variance of a double-precision floating-point strided array provided a known mean and using Neely's correction algorithm and alternative indexing semantics.
```
grammar

DhruvArvindSingh replied:
Resolved

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/include/stdlib/stats/base/dvarmpn.h:
```diff
@@ -31,7 +31,11 @@ extern "C" {
+/**
+* Computes the variance of a double-precision floating-point strided array provided a known mean and using Neely's correction algorithm using alternative indexing semantics.
+*/
+double API_SUFFIX( stdlib_strided_dvarmpn_ndarray )( const CBLAS_INT N, const double mean, const double correction, const double *X, const CBLAS_INT strideX, const CBLAS_INT offsetX );
```
```suggestion
double API_SUFFIX(stdlib_strided_dvarmpn_ndarray)( const CBLAS_INT N, const double mean, const double correction, const double *X, const CBLAS_INT strideX, const CBLAS_INT offsetX );
```
we don't give spacing in brackets when wrapping in `API_SUFFIX`

aayush0325 replied:
kindly check all other instances as well incase i miss any!

DhruvArvindSingh replied:
Resolved

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/include/stdlib/stats/base/dvarmpn.h:
```diff
@@ -31,7 +31,11 @@ extern "C" {
 * Computes the variance of a double-precision floating-point strided array provided a known mean and using Neely's correction algorithm.
 */
-double stdlib_strided_dvarmpn( const int64_t N, const double mean, const double correction, const double *X, const int64_t stride );
+double API_SUFFIX( stdlib_strided_dvarmpn )( const CBLAS_INT N, const double mean, const double correction, const double *X, const CBLAS_INT strideX );
```
```suggestion
double API_SUFFIX(stdlib_strided_dvarmpn)( const CBLAS_INT N, const double mean, const double correction, const double *X, const CBLAS_INT strideX );
```

DhruvArvindSingh replied:
Resolved

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/lib/dvarmpn.js:
```diff
@@ -32,7 +38,7 @@
 * @param {number} correction - degrees of freedom adjustment
 * @param {Float64Array} x - input array
-* @param {integer} stride - stride length
+* @param {integer} strideX - strideX length
```
```suggestion
* @param {integer} strideX - stride length
```

DhruvArvindSingh replied:
Resolved

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/lib/dvarmpn.js:
```diff
@@ -43,35 +49,9 @@
-	}
-	return (M2/n) - ((M/N)*(M/n));
+function dvarmpn( N, mean, correction, x, strideX ) {
+	var ox = stride2offset( N, strideX);
+	return ndarray( N, mean, correction, x, strideX, ox );
```
```suggestion
	return ndarray( N, mean, correction, x, strideX, stride2offset( N, strideX ) );
```
we only define a new variable for this when there are more than 1 offset parameters like in `stats/base/dcumax`. it doesn't really matter very much, just following what we've been doing in other packages here

DhruvArvindSingh replied:
Resolved

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/lib/dvarmpn.native.js:
```diff
@@ -28,11 +28,16 @@ var addon = require( './../src/addon.node' );
 /**
 * Computes the variance of a double-precision floating-point strided array provided a known mean and using Neely's correction algorithm.
 *
+* ## References
+*
+* -   Neely, Peter M. 1966. "Comparison of Several Algorithms for Computation of Means, Standard Deviations and Correlation Coefficients." _Communications of the ACM_ 9 (7). Association for Computing Machinery: 496–99. doi:[10.1145/365719.365958](https://doi.org/10.1145/365719.365958).
+* -   Schubert, Erich, and Michael Gertz. 2018. "Numerically Stable Parallel Computation of (Co-)Variance." In _Proceedings of the 30th International Conference on Scientific and Statistical Database Management_. New York, NY, USA: Association for Computing Machinery. doi:[10.1145/3221269.3223036](https://doi.org/10.1145/3221269.3223036).
+*
```
```suggestion
```
no need to add the reference if it wasn't previously there

DhruvArvindSingh replied:
Resolved

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/lib/dvarmpn.native.js:
```diff
@@ -28,11 +28,16 @@ var addon = require( './../src/addon.node' );
 * @param {number} correction - degrees of freedom adjustment
 * @param {Float64Array} x - input array
-* @param {integer} stride - stride length
+* @param {integer} strideX - strideX length
```
```suggestion
* @param {integer} strideX - stride length
```

DhruvArvindSingh replied:
Resolved

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/lib/ndarray.js:
```diff
@@ -32,21 +32,19 @@
 * @param {Float64Array} x - input array
-* @param {integer} stride - stride length
-* @param {NonNegativeInteger} offset - starting index
+* @param {integer} strideX - strideX length
```
```suggestion
* @param {integer} strideX - stride length
```

DhruvArvindSingh replied:
Resolved

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/src/main.c:
```diff
@@ -31,12 +32,34 @@
 * @param correction  degrees of freedom adjustment
 * @param X           input array
-* @param stride      stride length
+* @param strideX      strideX length
```
```suggestion
* @param strideX      stride length
```
kindly also adjust the spacing of the descriptions here. ideally, we would want them to be stacked over one another.

DhruvArvindSingh replied:
Resolved

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/src/main.c:
```diff
@@ -31,12 +32,34 @@
+}
+
+/**
+* Computes the variance of a double-precision floating-point strided array provided a known mean and using Neely's correction algorithm.
```
```suggestion
* Computes the variance of a double-precision floating-point strided array provided a known mean and using Neely's correction algorithm and alternative indexing semantics.
```

DhruvArvindSingh replied:
Resolved

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/src/main.c:
```diff
@@ -31,12 +32,34 @@
+*
+* ## References
+*
+* -   Neely, Peter M. 1966. "Comparison of Several Algorithms for Computation of Means, Standard Deviations and Correlation Coefficients." _Communications of the ACM_ 9 (7). Association for Computing Machinery: 496–99. doi:[10.1145/365719.365958](https://doi.org/10.1145/365719.365958).
+* -   Schubert, Erich, and Michael Gertz. 2018. "Numerically Stable Parallel Computation of (Co-)Variance." In _Proceedings of the 30th International Conference on Scientific and Statistical Database Management_. New York, NY, USA: Association for Computing Machinery. doi:[10.1145/3221269.3223036](https://doi.org/10.1145/3221269.3223036).
```
```suggestion
```
we don't need to add the reference twice.

DhruvArvindSingh replied:
Resolved

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/src/main.c:
```diff
@@ -31,12 +32,34 @@
+* @param mean        mean
+* @param correction  degrees of freedom adjustment
+* @param X           input array
+* @param strideX      strideX length
```
```suggestion
* @param strideX      stride length
```

aayush0325 replied:
kindly adjust the spacing here as well

DhruvArvindSingh replied:
Resolved

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/src/main.c:
```diff
@@ -31,12 +32,34 @@
+* @return            output value
+*/
+
+double API_SUFFIX( stdlib_strided_dvarmpn_ndarray )( const CBLAS_INT N, const double mean, const double correction, const double *X, const CBLAS_INT strideX, const CBLAS_INT offsetX ) {
```
```suggestion
double API_SUFFIX(stdlib_strided_dvarmpn_ndarray)( const CBLAS_INT N, const double mean, const double correction, const double *X, const CBLAS_INT strideX, const CBLAS_INT offsetX ) {
```

DhruvArvindSingh replied:
Resolved

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/src/main.c:
```diff
@@ -31,12 +32,34 @@
-double stdlib_strided_dvarmpn( const int64_t N, const double mean, const double correction, const double *X, const int64_t stride ) {
-	int64_t ix;
-	int64_t i;
+double API_SUFFIX( stdlib_strided_dvarmpn )( const CBLAS_INT N, const double mean, const double correction, const double *X, const CBLAS_INT strideX ) {
```
```suggestion
double API_SUFFIX(stdlib_strided_dvarmpn)( const CBLAS_INT N, const double mean, const double correction, const double *X, const CBLAS_INT strideX ) {
```

DhruvArvindSingh replied:
Resolved

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/lib/ndarray.native.js:
```diff
@@ -33,27 +32,20 @@ var addon = require( './dvarmpn.native.js' );
 * @param {Float64Array} x - input array
-* @param {integer} stride - stride length
-* @param {NonNegativeInteger} offset - starting index
+* @param {integer} strideX - strideX length
```
```suggestion
* @param {integer} strideX - stride length
```

DhruvArvindSingh replied:
Resolved

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/src/addon.c:
```diff
@@ -35,12 +35,31 @@
-	STDLIB_NAPI_ARGV_STRIDED_FLOAT64ARRAY( env, X, N, stride, argv, 3 );
-	STDLIB_NAPI_CREATE_DOUBLE( env, stdlib_strided_dvarmpn( N, mean, correction, X, stride ), v );
+	STDLIB_NAPI_ARGV_STRIDED_FLOAT64ARRAY( env, X, N, strideX, argv, 3 );
+	STDLIB_NAPI_CREATE_DOUBLE( env, stdlib_strided_dvarmpn( N, mean, correction, X, strideX ), v );
```
```suggestion
	STDLIB_NAPI_CREATE_DOUBLE( env, API_SUFFIX(stdlib_strided_dvarmpn)( N, mean, correction, X, strideX ), v );
```

you'll also need to include `stdlib/blas/base/shared.h`

DhruvArvindSingh replied:
Please verify whether we need `shared.h` package  is needed here, as it runs without it also

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/src/addon.c:
```diff
@@ -35,12 +35,31 @@
+	STDLIB_NAPI_ARGV_DOUBLE( env, mean, argv, 1 );
+	STDLIB_NAPI_ARGV_DOUBLE( env, correction, argv, 2 );
+	STDLIB_NAPI_ARGV_STRIDED_FLOAT64ARRAY( env, X, N, strideX, argv, 3 );
+	STDLIB_NAPI_CREATE_DOUBLE( env, stdlib_strided_dvarmpn_ndarray( N, mean, correction, X, strideX, offsetX ), v );
```
```suggestion
	STDLIB_NAPI_CREATE_DOUBLE( env, API_SUFFIX(stdlib_strided_dvarmpn_ndarray)( N, mean, correction, X, strideX, offsetX ), v );
```

DhruvArvindSingh replied:
Resolved

aayush0325 reviewed (changes requested):
also, you haven't made the required changes in `README.md` please refer to some of the previously merged PRs to get an idea of what changes you need to make there.

DhruvArvindSingh committed:
chore: resolved issues in stats/base/dvarmpn

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: passed
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: passed
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

DhruvArvindSingh committed:
Merge branch 'stdlib-js:develop' into ndarray/dvarmpn

DhruvArvindSingh committed:
chore: updated examples js

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

DhruvArvindSingh committed:
chore: updated readme

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

DhruvArvindSingh commented:
Thanks @aayush0325  for reviewing my PR.... I have made the necessary

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/README.md:
````diff
@@ -117,9 +117,9 @@ The function has the following parameters:
+The `N` and `strideX` parameters determine which elements in `x` are accessed at runtime. For example, to compute the [variance][variance] of every other element in `x`,
 
 ```javascript
 var Float64Array = require( '@stdlib/array/float64' );
````
please update this example to remove the usage of `@stdlib/math/base/special/floor`

aayush0325 replied:
same comment for all other instances

aayush0325 replied:
https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/stats/base/smax/README.md?plain=1#L74


see how the examples are written here for reference!

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/benchmark/c/benchmark.length.c:
```diff
@@ -120,6 +121,40 @@ static double benchmark( int iterations, int len ) {
+static double benchmark2( int iterations, int len ) {
+	double elapsed;
+	double x[ len ];
+	float v;
```
```suggestion
	double v;
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/benchmark/c/benchmark.length.c:
```diff
@@ -120,6 +121,40 @@ static double benchmark( int iterations, int len ) {
+	for ( i = 0; i < len; i++ ) {
+		x[ i ] = ( rand_double() * 20000.0 ) - 10000.0;
+	}
+	v = 0.0f;
```
```suggestion
	v = 0.0;
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/benchmark/c/benchmark.length.c:
```diff
@@ -120,6 +121,40 @@ static double benchmark( int iterations, int len ) {
+	t = tic();
+	for ( i = 0; i < iterations; i++ ) {
+		// cppcheck-suppress uninitvar
+		v = stdlib_strided_dvarmpn_ndarray( len, 0.0, 1, x, 1, 1 );
```
```suggestion
		v = stdlib_strided_dvarmpn_ndarray( len, 0.0, 1, x, 1, 0 );
```
just following the convention here

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/examples/c/example.c:
```diff
@@ -17,21 +17,20 @@
 int main( void ) {
 	// Create a strided array:
-	double x[] = { 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0 };
+	const double x[] = { 1.0f, 2.0f, 3.0f, 4.0f, 5.0f, 6.0f, 7.0f, 8.0f };
```
```suggestion
	const double x[] = { 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0 };
```
only single precision floating-point values are suffixed with a `f`

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/lib/dvarmpn.native.js:
```diff
@@ -41,10 +41,10 @@ var addon = require( './../src/addon.node' );
 *
 * var v = dvarmpn( x.length, 1.0/3.0, 1, x, 1 );
-* // returns ~4.3333
+*  //returns ~4.3333
```
```suggestion
* // returns ~4.3333
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/lib/ndarray.js:
```diff
@@ -32,21 +32,19 @@
-* var v = dvarmpn( N, 1.25, 1, x, 2, 1 );
-* // returns 6.25
+* var v = dvarmpn( 4, 1.25, 1, x, 2, 1 );
+*  //returns 6.25
```
```suggestion
* // returns 6.25
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dvarmpn/lib/ndarray.native.js:
```diff
@@ -33,27 +32,20 @@ var addon = require( './dvarmpn.native.js' );
-* var v = dvarmpn( N, 1.25, 1, x, 2, 1 );
-* // returns 6.25
+* var v = dvarmpn( 4, 1.25, 1, x, 2, 1 );
+*  //returns 6.25
```
```suggestion
* // returns 6.25
```

kgryte commented:
Ref: https://github.com/stdlib-js/stdlib/pull/5030
